ttnn/cpp/ttnn/operations/moreh/moreh_sgd/device/moreh_sgd_device_operation.cpp

```diff
@@ -0,0 +1,119 @@
+
+    tt::operations::primary::check_tensor(params_in, "moreh_sgd", "params_in");
+    tt::operations::primary::check_tensor(grad, "moreh_sgd", "grad");
+
+    if (tensor_args.momentum_buffer_in) {
+        tt::operations::primary::check_tensor(*tensor_args.momentum_buffer_in, "moreh_sgd", "momentum_buffer_in");
+    }
+
+    const auto& params_out = tensor_args.output_tensors.at(0);
+
+    if (params_out.has_value()) {
+        tt::operations::primary::check_tensor(params_out.value(), "moreh_sgd", "params_out");
+    }
+
+    if (tensor_args.output_tensors.at(1).has_value()) {
+        tt::operations::primary::check_tensor(tensor_args.output_tensors.at(1).value(), "moreh_sgd", "momentum_buffer_out");
+    }
+}
+
+MorehSgdOperation::program_factory_t MorehSgdOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    // For now we litteraly don't care and return a single factory. Whatever
+    return ProgramFactory{};
+}
+
+void MorehSgdOperation::validate_on_program_cache_miss(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+void MorehSgdOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+MorehSgdOperation::shape_return_value_t MorehSgdOperation::compute_output_shapes(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto input_tensor_shape = tensor_args.param_in.get_shape();
+
+    return {input_tensor_shape, input_tensor_shape};
+};
+
+MorehSgdOperation::tensor_return_value_t MorehSgdOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto& output_shapes = compute_output_shapes(operation_attributes, tensor_args);
+    auto dtype = tensor_args.param_in.get_dtype();
+    Layout layout{Layout::TILE};
+    auto device = tensor_args.param_in.device();
+
+    std::vector<std::optional<Tensor>> ret;
+    auto output_mem_config = operation_attributes.param_out_mem_config.value_or(tensor_args.param_in.memory_config());
+    auto momentum_buffer_out_mem_config = operation_attributes.momentum_buffer_out_mem_config.value_or(tensor_args.param_in.memory_config());
+
+    if (tensor_args.output_tensors.at(0).has_value()) {
+        ret.push_back(tensor_args.output_tensors.at(0).value());
+    } else {
+        ret.push_back(create_device_tensor(output_shapes.at(0).value(), dtype, layout, device, output_mem_config));
+    }
+
+    if (tensor_args.output_tensors.at(1).has_value()) {
+        ret.push_back(tensor_args.output_tensors.at(1).value());
+    } else {
+        ret.push_back(create_device_tensor(output_shapes.at(1).value(), dtype, layout, device, momentum_buffer_out_mem_config));
+    }
```

Comment:
```suggestion
    if (tensor_args.output_tensors.at(1).has_value()) {
        ret.push_back(tensor_args.output_tensors.at(1).value());
    } else if (this->momentum != 0.0f) {
        ret.push_back(create_device_tensor(output_shapes.at(1).value(), dtype, layout, device, momentum_buffer_out_mem_config));
    } else {
       ret.push_back(std::nullopt);
    }
```
The second output may not be necessary depending on the momentum value